### PR TITLE
Enable the use of config-file for identitymapping creation

### PR DESCRIPTION
**Pull request by adamjohnson01**
### Description
Enables creating IAM Identity Mappings using the config file as requested in #4631

### Checklist
- [ ] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/ctl/create/iamidentitymapping.go
Comment (cPu1) [inline pkg/ctl/create/iamidentitymapping.go:71]: Minor nit:

```suggestion
			if err := mapping.Validate(); err != nil {
			    return err
			}
			if err := createIAMIdentityMapping(cmd, *mapping); err != nil {
				return err
			}
```

File: pkg/ctl/create/iamidentitymapping.go
Comment (cPu1) [inline pkg/ctl/create/iamidentitymapping.go:81]: Minor nit:

```suggestion
		if err := options.Validate(); err != nil {
			return err
		}
		if err := createIAMIdentityMapping(cmd, options); err != nil {
			return err
		}
```

**Review by cPu1 — CHANGES_REQUESTED** [review]
@adamjohnson01, thanks for the PR! It LGTM, but since it adds a new field to the `ClusterConfig` schema, we also need to handle creating IAM identity mappings as part of `eksctl create cluster`. Without it, the feature will be incomplete and we can't get it out. Can you work on adding that? You can use the task management code for `create cluster` as a reference, it should be straightforward.

**Comment by adamjohnson01** [comment]
> @adamjohnson01, thanks for the PR! It LGTM, but since it adds a new field to the `ClusterConfig` schema, we also need to handle creating IAM identity mappings as part of `eksctl create cluster`. Without it, the feature is incomplete and we can't get it out. Can you work on adding that? You can use the task management code for `create cluster` as a reference, it should be straightforward.

@cPu1, sure. I actually thought the same thing last night after I created the PR. I will add that as well.

**Comment by adamjohnson01** [comment]
#### Create cluster with IAM identity mappings

2022-06-27 23:16:51 [ℹ]  eksctl version 0.105.0-dev+43f30788.2022-06-27T23:00:21Z
2022-06-27 23:16:51 [ℹ]  using region eu-west-1
2022-06-27 23:16:51 [✔]  using existing VPC (vpc-********) and subnets (private:map[eu-west-1a:{subnet-**** eu-west-1a 10.10.92.0/24 0} eu-west-1b:{subnet-**** eu-west-1b 10.10.93.0/24 0} eu-west-1c:{subnet-**** eu-west-1c 10.10.94.0/24 0} eu-west-1d:{subnet-**** eu-west-1d 10.10.95.0/24 0}] public:map[])
2022-06-27 23:16:51 [!]  custom VPC/subnets will be used; if resulting cluster doesn't function as expected, make sure to review the configuration of VPC/subnets
2022-06-27 23:16:51 [ℹ]  using Kubernetes version 1.21
2022-06-27 23:16:51 [ℹ]  creating EKS cluster "maps" in "eu-west-1" region with
2022-06-27 23:16:51 [ℹ]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
2022-06-27 23:16:51 [ℹ]  will create a CloudFormation stack for cluster itself and 0 managed nodegroup stack(s)
2022-06-27 23:16:51 [ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=eu-west-1 --cluster=maps'
2022-06-27 23:16:51 [ℹ]  Kubernetes API endpoint access will use default of {publicAccess=true, privateAccess=false} for cluster "maps" in "eu-west-1"
2022-06-27 23:16:51 [ℹ]  CloudWatch logging will not be enabled for cluster "maps" in "eu-west-1"
2022-06-27 23:16:51 [ℹ]  you can enable it with 'eksctl utils update-cluster-logging --enable-types={SPECIFY-YOUR-LOG-TYPES-HERE (e.g. all)} --region=eu-west-1 --cluster=maps'
2022-06-27 23:16:51 [ℹ]
2 sequential tasks: { create cluster control plane "maps",
    4 sequential sub-tasks: {
        wait for control plane to become ready,
        associate IAM OIDC provider,
        2 sequential sub-tasks: {
            create IAM role for serviceaccount "kube-system/aws-node",
            create serviceaccount "kube-system/aws-node",
        },
        restart daemonset "kube-system/aws-node",
    }
}
2022-06-27 23:16:51 [ℹ]  building cluster stack "eksctl-maps-cluster"
2022-06-27 23:16:51 [ℹ]  deploying stack "eksctl-maps-cluster"
2022-06-27 23:17:21 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:17:51 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:18:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:19:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:20:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:21:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:22:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:23:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:24:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:25:52 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:26:53 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:27:53 [ℹ]  waiting for CloudFormation stack "eksctl-maps-cluster"
2022-06-27 23:29:54 [ℹ]  building iamserviceaccount stack "eksctl-maps-addon-iamserviceaccount-kube-system-aws-node"
2022-06-27 23:29:54 [ℹ]  deploying stack "eksctl-maps-addon-iamserviceaccount-kube-system-aws-node"
2022-06-27 23:29:54 [ℹ]  waiting for CloudFormation stack "eksctl-maps-addon-iamserviceaccount-kube-system-aws-node"
2022-06-27 23:30:24 [ℹ]  waiting for CloudFormation stack "eksctl-maps-addon-iamserviceaccount-kube-system-aws-node"
2022-06-27 23:30:24 [ℹ]  serviceaccount "kube-system/aws-node" already exists
2022-06-27 23:30:24 [ℹ]  updated serviceaccount "kube-system/aws-node"
2022-06-27 23:30:24 [ℹ]  daemonset "kube-system/aws-node" restarted
2022-06-27 23:30:24 [ℹ]  waiting for the control plane availability...
2022-06-27 23:30:25 [✔]  saved kubeconfig as "/home/user/.kube/config"
2022-06-27 23:30:25 [ℹ]  no tasks
2022-06-27 23:30:25 [✔]  all EKS cluster resources for "maps" have been created
2022-06-27 23:30:26 [ℹ]  kubectl command should work with "/home/user/.kube/config", try 'kubectl get nodes'
2022-06-27 23:30:26 [ℹ]  iamIdentityMappings configuration detected, updating the auth ConfigMap
2022-06-27 23:30:26 [ℹ]  checking arn arn:aws:iam::***********:role/myRole against entries in the auth ConfigMap
2022-06-27 23:30:26 [ℹ]  adding identity "arn:aws:iam::***********:role/myRole" to auth ConfigMap
2022-06-27 23:30:26 [ℹ]  adding account "***********" to auth ConfigMap
2022-06-27 23:30:26 [✔]  EKS cluster "maps" in "eu-west-1" region is ready

File: pkg/actions/iamidentitymapping/create.go
Comment (Himangini) [inline pkg/actions/iamidentitymapping/create.go:7]: we've moved to aws-sdk-go-v2 now, can you use v2 here?

**Review by Himangini — COMMENTED** [review]
One suggested change for aws-sdk-go-v2 otherwise PR lgtm 👍🏻

File: pkg/actions/iamidentitymapping/create.go
Comment (adamjohnson01) [reply pkg/actions/iamidentitymapping/create.go:7]: Will update it, thanks!

File: pkg/apis/eksctl.io/v1alpha5/identitymapping.go
Comment (cPu1) [inline pkg/apis/eksctl.io/v1alpha5/identitymapping.go:21]: Nit: superfluous parentheses. 
```suggestion
	return im.hasARN() && im.hasUsername() && im.hasGroups()
```

File: pkg/ctl/create/iamidentitymapping.go
Comment (cPu1) [inline pkg/ctl/create/iamidentitymapping.go:42]: Alternatively:

```suggestion
	var identityMapping api.IAMIdentityMapping
	cfg.IAMIdentityMappings = []*api.IAMIdentityMapping{&identityMapping}
	cmd.FlagSetGroup.InFlagSet("IAMIdentityMapping", func(fs *pflag.FlagSet) {
		fs.StringVar(&identityMapping.Account, "account", "", "Account ID to automatically map to its username")
		cmdutils.AddIAMIdentityMappingARNFlags(fs, cmd, &identityMapping.ARN, "create")
		fs.StringVar(&identityMapping.Username, "username", "", "User name within Kubernetes to map to IAM role")
		fs.StringSliceVar(&identityMapping.Groups, "group", []string{}, "Groups within Kubernetes to which IAM role is mapped")
		fs.StringVar(&identityMapping.ServiceName, "service-name", "", "Service name; valid value: emr-containers")
		fs.StringVar(&identityMapping.Namespace, "namespace", "", "Namespace in which to create RBAC resources (only valid with --service-name)")
		fs.BoolVar(&identityMapping.NoDuplicateArns, "no-duplicate-arns", false, "Throw error when an aws-auth record already exists with the given ARN")
	})
```

File: pkg/ctl/create/cluster.go
Comment (cPu1) [inline pkg/ctl/create/cluster.go:445]: This should go in a task, created in `CreateExtraClusterConfigTasks` (which has examples of other post-cluster-creation tasks). This can be appended as a task after the `IdentityProviders` task.

File: pkg/ctl/create/iamidentitymapping.go
Comment (cPu1) [inline pkg/ctl/create/iamidentitymapping.go:41]:
```suggestion
		fs.BoolVar(&cfg.IAMIdentityMappings[0].NoDuplicateArns, "no-duplicate-arns", false, "Throw error when an aws-auth record already exists with the given ARN.")
```

I know this wasn't introduced by your PR.

File: pkg/apis/eksctl.io/v1alpha5/identitymapping.go
Comment (cPu1) [inline pkg/apis/eksctl.io/v1alpha5/identitymapping.go:17]:
```suggestion
	NoDuplicateARNs bool     `json:"noDuplicateARNs,omitempty"`
```
This would also require renaming references to this field in the code and examples.

File: examples/35-iamidentitymappings.yaml
Comment (cPu1) [inline examples/35-iamidentitymappings.yaml:14]: Based on the suggestion below:
```suggestion
    noDuplicateARNs: true # prevents shadowing of ARNs
```

File: examples/35-iamidentitymappings.yaml
Comment (cPu1) [inline examples/35-iamidentitymappings.yaml:18]:
```suggestion
    noDuplicateARNs: true # prevents shadowing of ARNs
```

**Review by cPu1 — CHANGES_REQUESTED** [review]
Thanks for the contribution! :tada:  I have left some comments but otherwise LGTM. We can merge it as soon as the requested changes are addressed.

File: pkg/ctl/create/cluster.go
Comment (adamjohnson01) [reply pkg/ctl/create/cluster.go:445]: @cPu1, that function has the following comment `// CreateExtraClusterConfigTasks returns all tasks for updating cluster configuration not depending on the control plane availability`. This updates the `aws-auth`  configmap so it does depend on control plane availability.

Comment (cPu1) [reply pkg/ctl/create/cluster.go:445]: That comment is outdated. The first task created in `CreateExtraClusterConfigTasks` waits for the control plane availability, so subsequent tasks can issue requests to the API server. Please feel free to update that comment, or I can do it later.

Comment (Himangini) [reply pkg/ctl/create/cluster.go:445]: Let's update the comment while we are here :)

Comment (adamjohnson01) [reply pkg/ctl/create/cluster.go]: @cPu1 and @Himangini , I have updated the comment and converted that code to be a task.

**Comment by adamjohnson01** [comment]
> Thanks for the contribution! 🎉 I have left some comments but otherwise LGTM. We can merge it as soon as the requested changes are addressed.

@cPu1 , all the changes have been addressed. Let me know if there is anything else.

**Comment by cPu1** [comment]
> > Thanks for the contribution! tada I have left some comments but otherwise LGTM. We can merge it as soon as the requested changes are addressed.
> 
> @cPu1 , all the changes have been addressed. Let me know if there is anything else.

Thanks @adamjohnson01. I skimmed through the changes and it LGTM. We are planning a code-freeze to get some features out this week, so we'll be able to merge your PR next week.

**Comment by adamjohnson01** [comment]
> > > Thanks for the contribution! tada I have left some comments but otherwise LGTM. We can merge it as soon as the requested changes are addressed.
> > 
> > 
> > @cPu1 , all the changes have been addressed. Let me know if there is anything else.
> 
> Thanks @adamjohnson01. I skimmed through the changes and it LGTM. We are planning a code-freeze to get some features out this week, so we'll be able to merge your PR next week.

@cPu1, will this be merged soon? We are keen to start using the functionality.

File: pkg/eks/tasks.go
Comment (cPu1) [inline pkg/eks/tasks.go:287]:
```suggestion
			Description: "create IAM identity mappings",
```

File: pkg/eks/tasks.go
Comment (cPu1) [inline pkg/eks/tasks.go:300]:
```suggestion
					return errors.Wrap(err, "error initialising iamidentitymapping")
```

File: pkg/eks/tasks.go
Comment (cPu1) [inline pkg/eks/tasks.go:304]: This validation should happen sooner, before any tasks are run. Can you move this to `ValidateClusterConfig`?

**Review by cPu1 — COMMENTED** [review]
>@cPu1, will this be merged soon? We are keen to start using the functionality.

Apologies for the delay. I have left some comments on the new changes. We should be good to merge after the comments are addressed.

**Comment by adamjohnson01** [comment]
@cPu1, I have made all of the requested changes. Let me know if there is anything else.

**Comment by jacklei** [comment]
@cPu1 bump

**Review by Himangini — APPROVED** [review]
LGTM all the requested changes have been addressed 🎉

File: pkg/apis/eksctl.io/v1alpha5/validation.go
Comment (cPu1) [inline pkg/apis/eksctl.io/v1alpha5/validation.go:1395]: This isn't required. 

```suggestion
```

**Review by cPu1 — APPROVED** [review]

**Comment by adamjohnson01** [comment]
> @adamjohnson01, apologies for the delay in getting this merged. Thanks for the contribution!

@cPu1, no problem! 🙂

**Pull request merged**
